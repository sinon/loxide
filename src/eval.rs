//! Eval module
//!
//! Responsible for evlauting the AST and returning the computed values
//!

use std::fmt::Display;

use crate::parser::{Expr, LiteralAtom, Parser};

pub enum EvaluatedValue<'de> {
    String(&'de str),
    Number(f64),
    Nil,
    Bool(bool),
}

impl Display for EvaluatedValue<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            EvaluatedValue::String(s) => write!(f, "{s}"),
            EvaluatedValue::Number(n) => write!(f, "{n}"),
            EvaluatedValue::Nil => write!(f, "nil"),
            EvaluatedValue::Bool(b) => write!(f, "{b:}"),
        }
    }
}

pub struct Eval<'de> {
    parser: Parser<'de>,
}

impl<'de> Eval<'de> {
    pub fn new(input: &'de str) -> Self {
        Eval {
            parser: Parser::new(input),
        }
    }
}

impl<'de> Iterator for Eval<'de> {
    type Item = Result<EvaluatedValue<'de>, String>;

    fn next(&mut self) -> Option<Self::Item> {
        let expr = self.parser.next()?;

        match expr {
            Ok(e) => return self.evaluate_expression(e),
            Err(_) => todo!(),
        }
        todo!()
    }
}
impl<'de> Eval<'de> {
    fn evaluate_expression(&self, expr: Expr<'de>) -> Option<Result<EvaluatedValue<'de>, String>> {
        match expr {
            Expr::Binary {
                left,
                operator,
                right,
            } => {
                todo!()
            }
            Expr::Unary { operator, right } => todo!(),
            Expr::Literal(literal_atom) => match literal_atom {
                LiteralAtom::String(s) => {
                    return Some(Ok(EvaluatedValue::String(s)));
                }
                LiteralAtom::Number(num) => return Some(Ok(EvaluatedValue::Number(num))),
                LiteralAtom::Nil => return Some(Ok(EvaluatedValue::Nil)),
                LiteralAtom::Bool(b) => {
                    return Some(Ok(EvaluatedValue::Bool(b)));
                }
            },
            Expr::Grouping(expr) => self.evaluate_expression(*expr),
        }
    }
}
