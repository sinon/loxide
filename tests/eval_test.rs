use insta::{assert_debug_snapshot, assert_snapshot};
use loxide::eval::Eval;
use rstest::*;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        settings.set_description(format!($($expr,)*));
        settings.set_omit_expression(true);
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
#[case("true")]
#[case("false")]
#[case("nil")]
#[case("\"hello world\"")]
#[case("10.4")]
#[case("10")]
#[case("10.0")]
#[case("(\"hello world!\")")]
#[case("(10.5)")]
#[case("((false))")]
#[case("-73")]
#[case("!true")]
#[case("!nil")]
#[case("!10.40")]
#[case("!((false))")]
#[case("42 / 5")]
#[case("18 * 3 / (3 * 6)")]
#[case("(10.40 * 2) / 2")]
#[case("23 + 28 - (-(61 - 99))")]
#[case("\"foo\" + \"bar\"")]
#[case("\"42\" + \"24\"")]
#[case("57 > -65")]
#[case("11 >= 11")]
#[case("(54 - 67) >= -(114 / 57 + 11)")]
#[case("\"hello\" == \"world\"")]
#[case("\"foo\" != \"bar\"")]
#[case("\"foo\" == \"foo\"")]
#[case("61 == \"61\"")]
#[case("61 != \"61\"")]
#[case("\"61\" != 61")]
fn test_eval_literals(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let exprs: Vec<String> = Eval::new(input)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    assert_snapshot!(exprs.join("\n"));
}

#[rstest]
#[case("-\"hello world!\"")]
#[case("-true")]
#[case("-(\"world\" + \"foo\")")]
#[case("\"foo\" * 42")]
#[case("true / 2")]
#[case("(\"foo\" * \"bar\")")]
#[case("false / true")]
#[case("\"foo\" + true")]
#[case("42 - true")]
#[case("true + false")]
#[case("\"foo\" - \"bar\"")]
#[case("\"foo\" < false")]
#[case("true < 2")]
#[case("(\"foo\" + \"bar\") < 42")]
#[case("false > true")]
fn test_eval_errors(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let mut errors = Vec::new();
    for e in Eval::new(input) {
        match e {
            Ok(_) => {}
            Err(e) => errors.push(e),
        }
    }
    assert_debug_snapshot!(errors);
}
