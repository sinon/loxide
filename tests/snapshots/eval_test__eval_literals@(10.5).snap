---
source: tests/eval_test.rs
description: (10.5)
---
10.5
